### PR TITLE
supercluster: added initial,map,reduce to Options

### PR DESCRIPTION
**Pull request by kl4kennylee81**
Talked with @DenisCarriere in #21211 

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes:
definition of initial,map,reduce in the Options for supercluster
https://www.npmjs.com/package/@types/supercluster

source code with usage of map,reduce,initial
https://github.com/mapbox/supercluster/blob/master/index.js

This is my first time opening an open pull request in a public repository. Let me know if there's anything I forgot to do. Thanks!

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@kl4kennylee81 Please fix the failures indicated in the Travis CI log.

**Comment by DenisCarriere** [comment]
I'm going to look at this PR over the weekend, first will be to fix the tests which are mostly linting issues.

File: types/supercluster/index.d.ts
Comment (DenisCarriere) [inline types/supercluster/index.d.ts]: @kl4kennylee81 What happened with the indentation?

image

File: types/supercluster/index.d.ts
Comment (DenisCarriere) [inline types/supercluster/index.d.ts]: It would be best to define a defaults for both `MapProps` & `ReduceProps`

```ts
export default function supercluster<MapProps extends any, ReduceProps extends any>(
```

File: types/supercluster/index.d.ts
Comment (kl4kennylee81) [reply types/supercluster/index.d.ts]: sorry my text editor had different tab/space setting and I committed without looking it over closely. My apologies.

File: types/supercluster/index.d.ts
Comment (kl4kennylee81) [reply types/supercluster/index.d.ts]: you're definitely right have defined defaults makes it easier for the users of the library to not have to define it explicitly if they don't need it. Thanks for making the suggestion.

so turns out after running the linter it was complaining about
'''
ERROR: 97:20  no-unnecessary-generics  Type parameter MapProps is used only once. See: https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md
ERROR: 97:30  no-unnecessary-generics  Type parameter ReduceProps is used only once. See: https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md
'''

**Comment by kl4kennylee81** [comment]
@DenisCarriere Apologize about the linter issue I pushed it out without looking over the commit. Sorry about that

File: types/supercluster/index.d.ts
Comment (DenisCarriere) [inline types/supercluster/index.d.ts]: Typically I try to use the param names as the original source code ([`supercluster/index.js`](https://github.com/mapbox/supercluster/blob/master/index.js#L26-L33)), for this definition the params would be:

- `acc` => `accumulated`
- `geoJsonProperties` => `props`
- `value` => `props`

or pull the param names from the examples in the main README page.

```js
var index = supercluster({
    initial: function() { return {sum: 0}; },
    map: function(props) { return {sum: props.myValue}; },
    reduce: function(accumulated, props) { accumulated.sum += props.sum; }
});
```

File: types/supercluster/supercluster-tests.ts
Comment (DenisCarriere) [inline types/supercluster/supercluster-tests.ts]: I'd simplify these test examples since you are now using `any` and I'd try to avoid overly complicated Typescript test cases.


```ts
const initialFunction = () => {
  return 0
}
```


```ts
const mapFunction = (props) => {
  return props
}
```

```ts
const reduceFunction = (accumulated, props) => {
    if (props.name === 'Ashley') accumulated++; 
    return accumulated;
};
```

File: types/supercluster/supercluster-tests.ts
Comment (DenisCarriere) [inline types/supercluster/supercluster-tests.ts]: No longer needed since we are using `any` to define the properties.

**Review by DenisCarriere — CHANGES_REQUESTED** [review]
Review comments

File: types/supercluster/index.d.ts
Comment (DenisCarriere) [inline types/supercluster/index.d.ts]: Why not use the same wording as the source code ([`supercluster/index.js`](https://github.com/mapbox/supercluster/blob/master/index.js#L26-L33))

```ts
/**
 * a reduce function for calculating custom cluster properties
 *
 * @example
 * function (accumulated, props) { accumulated.sum += props.sum; }
 */
reduce?: (accumulated: any, props: any) => void;

/**
 * initial properties of a cluster (before running the reducer)
 *
 * @example
 * function () { return {sum: 0}; }
 */
initial?: () => any;

/**
 * properties to use for individual points when running the reducer
 *
 * @example
 * function (props) { return {sum: props.my_value}; }
 */
map?: (props: any) => any;
```

File: types/supercluster/index.d.ts
Comment (kl4kennylee81) [reply types/supercluster/index.d.ts]: yea that makes sense will make the change

File: types/supercluster/supercluster-tests.ts
Comment (kl4kennylee81) [reply types/supercluster/supercluster-tests.ts]: yea will change the test's interface declaration

File: types/supercluster/supercluster-tests.ts
Comment (kl4kennylee81) [reply types/supercluster/supercluster-tests.ts]: sure I'll just do the sum accumulated

File: types/supercluster/index.d.ts
Comment (kl4kennylee81) [reply types/supercluster/index.d.ts]: will change to the comment from the source code thanks!

**Comment by kl4kennylee81** [comment]
@DenisCarriere updated based on your comments appreciate the thorough feedback. Thanks for looking it over. let me know if it's good now.

**Comment by RyanCavanaugh** [comment]
@DenisCarriere - Thanks for your review of this PR! Can you please look at the new code and update your review status if appropriate?

**Review by DenisCarriere — APPROVED** [review]
👍 Looks great! Ready to be merged

**Comment by kl4kennylee81** [comment]
@DenisCarriere thanks alot for the help. Also @RyanCavanaugh thanks for checking up as well. It's my first time doing this so really appreciate the assistance.

**Comment by RyanCavanaugh** [comment]
Approved by a listed owner. PR ready to merge pending express review by a maintainer.

**Pull request merged**
